File: android/app/src/main/java/com/bwqr/mavinote/viewmodels/AccountViewModel.kt
package com.bwqr.mavinote.viewmodels

import com.bwqr.mavinote.models.Account
import com.bwqr.mavinote.models.Device
import com.bwqr.mavinote.models.Mavinote
import com.bwqr.mavinote.reax.DeBool
import com.bwqr.mavinote.reax.DeList
import com.bwqr.mavinote.reax.DeOption
import com.bwqr.mavinote.reax.DeString
import com.bwqr.mavinote.reax.DeUnit
import com.bwqr.mavinote.reax.Runtime
import kotlinx.coroutines.flow.Flow

class AccountViewModel {
    companion object {
        fun accounts(): Flow<List<Account>> =
            Runtime.runStream(DeList(Account)) { _accounts(it) }

        suspend fun account(accountId: Int): Account? =
            Runtime.runOnce(DeOption(Account)) { _account(it, accountId) }

        suspend fun devices(accountId: Int): List<Device> =
            Runtime.runOnce(DeList(Device)) { _devices(it, accountId) }

        suspend fun addDevice(accountId: Int, fingerprint: String): Unit =
            Runtime.runOnceUnit { _addDevice(it, accountId, fingerprint) }

        suspend fun deleteDevice(accountId: Int, deviceId: Int) =
            Runtime.runOnceUnit { _deleteDevice(it, accountId, deviceId) }

        suspend fun requestVerification(email: String): String =
            Runtime.runOnce(DeString) { _requestVerification(it, email) }

        suspend fun waitVerification(token: String) =
            Runtime.runOnceUnit { _waitVerification(it, token) }

        suspend fun sendVerificationCode(email: String) =
            Runtime.runOnceUnit { _sendVerificationCode(it, email) }

        suspend fun signUp(email: String, code: String) =
            Runtime.runOnceUnit { _signUp(it, email, code) }

        suspend fun mavinoteAccount(accountId: Int): Mavinote? =
            Runtime.runOnce(DeOption(Mavinote)) { _mavinoteAccount(it, accountId) }

        suspend fun addAccount(email: String) = Runtime.runOnceUnit { _addAccount(it, email) }

        suspend fun removeAccount(accountId: Int) =
            Runtime.runOnceUnit { _removeAccount(it, accountId) }

        suspend fun sendAccountCloseCode(accountId: Int) =
            Runtime.runOnceUnit { _sendAccountCloseCode(it, accountId) }

        suspend fun closeAccount(accountId: Int, code: String) =
            Runtime.runOnceUnit { _closeAccount(it, accountId, code) }

        suspend fun publicKey(): String = Runtime.runOnce(DeString) { _publicKey(it) }

        fun listenNotifications(accountId: Int): Flow<Unit> = Runtime.runStream(DeUnit) { _listenNotifications(it, accountId) }

        suspend fun welcomeShown(): Boolean = Runtime.runOnce(DeBool) { _welcomeShown(it) }

        suspend fun updateWelcomeShown(shown: Boolean): Unit = Runtime.runOnceUnit { _updateWelcomeShown(it, shown) }
    }
}

private external fun _accounts(streamId: Int): Long
private external fun _account(onceId: Int, accountId: Int): Long
private external fun _mavinoteAccount(onceId: Int, accountId: Int): Long
private external fun _devices(onceId: Int, accountId: Int): Long
private external fun _addDevice(onceId: Int, accountId: Int, fingerprint: String): Long
private external fun _deleteDevice(onceId: Int, accountId: Int, deviceId: Int): Long
private external fun _requestVerification(onceId: Int, email: String): Long
private external fun _waitVerification(onceId: Int, token: String): Long
private external fun _sendVerificationCode(onceId: Int, email: String): Long
private external fun _signUp(onceId: Int, email: String, code: String): Long
private external fun _addAccount(onceId: Int, email: String): Long
private external fun _removeAccount(onceId: Int, accountId: Int): Long
private external fun _sendAccountCloseCode(onceId: Int, accountId: Int): Long
private external fun _closeAccount(onceId: Int, accountId: Int, code: String): Long
private external fun _publicKey(onceId: Int): Long
private external fun _listenNotifications(streamId: Int, accountId: Int): Long
private external fun _welcomeShown(onceId: Int): Long
private external fun _updateWelcomeShown(onceId: Int, shown: Boolean): Long

File: ios/Mavinote/ViewModels/AccountViewModel.swift
typealias AccountResult<T> = Result<T, NoteError>

class AccountViewModel {
    static func accounts() -> AsyncStream<AccountResult<[Account]>> {
        return Runtime.runStream { reax_account_accounts($0) }
    }

    static func account(_ accountId: Int32) async -> AccountResult<Account?> {
        return await Runtime.runOnce { reax_account_account($0, accountId) }
    }

    static func devices(_ accountId: Int32) async -> AccountResult<[Device]> {
        return await Runtime.runOnce { reax_account_devices($0, accountId) }
    }

    static func addDevice(_ accountId: Int32, _ fingerprint: String) async -> AccountResult<()> {
        return await Runtime.runOnceUnit { reax_account_add_device($0, accountId, fingerprint) }
    }

    static func deleteDevice(_ accountId: Int32, _ deviceId: Int32) async -> AccountResult<()> {
        return await Runtime.runOnceUnit { reax_account_delete_device($0, accountId, deviceId) }
    }

    static func requestVerification(_ email: String) async -> AccountResult<String> {
        return await Runtime.runOnce { reax_account_request_verification($0, email) }
    }

    static func waitVerification(_ token: String) async -> AccountResult<()> {
        return await Runtime.runOnceUnit { reax_account_wait_verification($0, token) }
    }

    static func sendVerificationCode(_ email: String) async -> AccountResult<()> {
        return await Runtime.runOnceUnit { reax_account_send_verification_code($0, email) }
    }

    static func signUp(_ email: String, _ code: String) async -> AccountResult<()> {
        return await Runtime.runOnceUnit { reax_account_sign_up($0, email, code) }
    }

    static func mavinoteAccount(_ accountId: Int32) async -> AccountResult<Mavinote?> {
        return await Runtime.runOnce { reax_account_mavinote_account($0, accountId) }
    }

    static func addAccount(_ email: String) async -> AccountResult<()> {
        return await Runtime.runOnceUnit { reax_account_add_account($0, email) }
    }

    static func removeAccount(_ accountId: Int32) async -> AccountResult<()> {
        return await Runtime.runOnceUnit { reax_account_remove_account($0, accountId) }
    }

    static func sendAccountCloseCode(_ accountId: Int32) async -> AccountResult<()> {
        return await Runtime.runOnceUnit { reax_account_send_account_close_code($0, accountId) }
    }

    static func closeAccount(_ accountId: Int32, _ code: String) async -> AccountResult<()> {
        return await Runtime.runOnceUnit { reax_account_close_account($0, accountId, code) }
    }

    static func publicKey() async -> AccountResult<String> {
        return await Runtime.runOnce { reax_account_public_key($0) }
    }

    static func listenNotifications(_ accountId: Int32) -> AsyncStream<AccountResult<DeUnit>> {
        return Runtime.runStream { reax_account_listen_notifications($0, accountId) }
    }

    static func welcomeShown() async -> AccountResult<Bool> {
        return await Runtime.runOnce { reax_account_welcome_shown($0) }
    }

    static func updateWelcomeShown(_ shown: Bool) async -> AccountResult<DeUnit> {
        return await Runtime.runOnce { reax_account_update_welcome_shown($0, shown) }
    }
}
